docs/spec.md

```diff
@@ -140,240 +43,355 @@ stablehlo.func @main(
 }
 ```
 
-## Execution
+### Functions
 
-### Sequential execution
+```ebnf
+Func        ::= 'stablehlo' '.' 'func' FuncId FuncInputs FuncOutputs '{' FuncBody '}'
+FuncInputs  ::= '(' [FuncInput {',' FuncInput}] `)`
+FuncInput   ::= '%' ValueId ':' ValueType
+FuncOutputs ::= ['->' FuncOutput, {',' FuncOutput}]
+FuncOutput  ::= ValueType
+FuncBody    ::= {Op}
+```
 
-A StableHLO program is executed by providing input values to the `main` function
-and computing output values. Output values of a function are computed by
-executing the graph of ops rooted in the corresponding `return` op.
+**StableHLO functions** (which are also called **named functions**) have
+an identifier, inputs/outputs and a body. In the future, we are planning to
+introduce additional metadata for functions to achieve better compatibility
+with HLO ([#425](https://github.com/openxla/stablehlo/issues/425)).
 
-The execution order is implementation-defined, as long as ops are executed
-before their uses. Possible execution orders of the example program above are
-`%0` → `%1` → `%2` → `%3` → `%4` → `return` or `%3` → `%0` → `%1` → `%2` → `%4`
-→ `return`.
+### Identifiers
 
-More formally, a **StableHLO process** is a combination of:
-(1) a StableHLO program, (2) operation statuses (not executed yet,
-already executed), and (3) intermediate values that the process is working on.
-The process starts with input values to the `main` function, progresses through
-the graph of ops updating operation statuses and intermediate values and
-finishes with output values. Further formalization is TBD.
+```ebnf
+FuncId  ::= '@' letter {letter | digit}
```

Comment:
This is not enforced currently. We currently support MLIR identifiers: https://mlir.llvm.org/docs/LangRef/#identifiers-and-keywords

```
  stablehlo.func @func.with.dot() {
    stablehlo.return
  }
```

Reply:
This section may be easier to write if we define an "identifier" production of our own